server/src/domain/asset/asset.service.ts

```diff
@@ -32,4 +45,102 @@ export class AssetService {
 
     return Promise.all(requests).then((results) => results.filter((result) => result.assets.length > 0));
   }
+
+  async downloadFile(authUser: AuthUserDto, id: string): Promise<ImmichReadStream> {
+    await this.access.requirePermission(authUser, Permission.ASSET_DOWNLOAD, id);
+
+    const [asset] = await this.assetRepository.getByIds([id]);
+    if (!asset) {
+      throw new BadRequestException('Asset not found');
+    }
+
+    return this.storageRepository.createReadStream(asset.originalPath, asset.mimeType);
+  }
+
+  async getDownloadInfo(authUser: AuthUserDto, dto: DownloadDto): Promise<DownloadResponseDto> {
+    const targetSize = dto.archiveSize || HumanReadableSize.GiB * 4;
+    const archives: DownloadArchiveInfo[] = [];
+    let archive: DownloadArchiveInfo = { size: 0, assetIds: [] };
+
+    const assetPagination = await this.getDownloadAssets(authUser, dto);
+    for await (const assets of assetPagination) {
+      // motion part of live photos
+      const motionIds = assets.map((asset) => asset.livePhotoVideoId).filter<string>((id): id is string => !!id);
+      if (motionIds.length > 0) {
+        assets.push(...(await this.assetRepository.getByIds(motionIds)));
+      }
+
+      for (const asset of assets) {
+        archive.size += Number(asset.exifInfo?.fileSizeInByte || 0);
+        archive.assetIds.push(asset.id);
+
+        if (archive.size > targetSize) {
+          archives.push(archive);
+          archive = { size: 0, assetIds: [] };
+        }
+      }
+
+      if (archive.assetIds.length > 0) {
+        archives.push(archive);
+      }
+    }
+
+    return {
+      totalSize: archives.reduce((total, item) => (total += item.size), 0),
+      archives,
+    };
+  }
+
+  async downloadArchive(authUser: AuthUserDto, dto: AssetIdsDto): Promise<ImmichReadStream> {
+    await this.access.requirePermission(authUser, Permission.ASSET_DOWNLOAD, dto.assetIds);
+
+    const zip = this.storageRepository.createZipStream();
+    const assets = await this.assetRepository.getByIds(dto.assetIds);
+    const paths: Record<string, boolean> = {};
+
+    for (const { originalPath, originalFileName } of assets) {
+      const ext = extname(originalPath);
+      let filename = `${originalFileName}${ext}`;
+      for (let i = 0; i < 10_000; i++) {
+        if (!paths[filename]) {
+          break;
+        }
+        filename = `${originalFileName}+${i + 1}${ext}`;
+      }
+
+      paths[filename] = true;
+      zip.addFile(originalPath, filename);
+    }
```

Comment:
Just because it was merged doesn't mean it can't be fixed anymore lol :stuck_out_tongue: 

Do you have an example of how array.group would look? I've never used it before.

Reply:
group is still experimental, so I can implement the suggestion from @brighteyed instead.